### PR TITLE
Library data

### PR DESCRIPTION
**Pull request by adrian-ponce-wolox**
## Summary ##
**Trello card**: https://trello.com/c/oBIzM5n0/13-vista-de-biblioteca
**Trello card**: https://trello.com/c/SvEAkxWU/14-tabla-paginada-de-libros

Request book list from API and used data to show booktableview with name, author and image for each book cell.

## Test Plan ##

  * Open the app
  * Login with google
  * The main tab bar view should appear after a successfull login.
  * Book table should be displayed after a while

### iPhone 7s
login-view
libraryview

## Screencast ##

libraryview

### CODE REVIEW

File: WBooks/BookTable/BookTableViewCell.swift [hunk 1]
```diff
@@ -21,4 +28,34 @@ final class BookTableViewCell: UITableViewCell {
     static var ShadowOffset: CGSize  = CGSize(width: -1, height: 1)
     static var ShadowOpacity: Float32  = 0.1
     
+    func bindViewModel(viewModel: BookViewModel) {
+        containerView.layer.cornerRadius = BookTableViewCell.CornerRadius
```
Comment (danielaRiesgo) [inline WBooks/BookTable/BookTableViewCell.swift]: These 3 are something that is always the same, and does not depend on the `viewModel ` received, so it shouldn't be done here. (Remember the `didSet` property in `var`s ?)

File: WBooks/BookTable/BookTableViewCell.swift [hunk 2]
```diff
@@ -21,4 +28,34 @@ final class BookTableViewCell: UITableViewCell {
...
+        
+        authorLabel.text = viewModel.author
+        titleLabel.text = viewModel.title
+        
+        if let imageUrl = viewModel.imageURL {
+            coverImageView.rex_image <~ imageProducer(imageUrl).map { return $0 }.liftError()
+        }
+    }
+    
+    func imageProducer(imageUrl: NSURL) -> SignalProducer<UIImage, ImageFetcherError> {
```
Comment (danielaRiesgo) [inline WBooks/BookTable/BookTableViewCell.swift]: Once more, improve the name of this function as said for `bookListProducer`.

File: WBooks/BookTable/BookViewModel.swift [hunk 3]
```diff
@@ -28,8 +32,23 @@ extension BookViewModel {
         return _book.author
     }
     
-    var coverImageURL: NSURL {
-        return NSURL(string: _book.image)!
+    var imageURL: NSURL? {        
+        if let urlString = _book.image_url {
```
Comment (danielaRiesgo) [inline WBooks/BookTable/BookViewModel.swift]: This is the same as doing:
`return  _book.image_url.flatMap { NSURL(string: $0) }`
You can see the difference between `map` and `flatMap` in optionals (and all optionals-related swift syntax in the From ObjC to Swift 2.3 presentation).

File: WBooks/Models/BooksList.swift [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import Argo
+import Curry
+
+public struct BooksList {
+    let books: [Book]
+}
+
+extension BooksList: Decodable {
+        
+    public static func decode(json: JSON) -> Decoded<BooksList> {
```
Comment (danielaRiesgo) [inline WBooks/Models/BooksList.swift]: All this is not needed !
Argo already knows how to decode an array of books ! (if `Book` is a `Decodable` type).
Just do:
```
return curry(BooksList.init)
    <^> json <|| "books"
```
BUT this works if the json is something like
```
{
    "books" : [ ...array of books... ]
}
```

If the json is directly the array, then don't make this class decodable !! Argo already knows how to decode a list of Books (because Book is a Decodable type). So let Argo decode it, and then, if you want, you can create a BookList directly with the array of books. (to decode the array, you have the decode function for free, so you will be able to do `decode(JSON).toResult()` directly).

The other question is if this type is necessary or you could just use an array of books. Since now I don't see it having any other use than to hold a list of books, it seems redundant, but if you think this will grow, and you will have many functions in this class to process that books, then leave it. It doesn't seem to me it's necessary, because an array already has a great collection of useful functions to process it.

File: WBooks/BookTab/HomeTabViewController.swift [hunk 5]
```diff
@@ -66,21 +62,10 @@ private extension HomeTabViewController {
...
-                    alert.addAction(UIAlertAction(title: "OK".localized(), style: UIAlertActionStyle.Default, handler: .None))
-                    self.presentViewController(alert, animated: true, completion: .None)
-                }
-            }
-
-        }
+        let booksRepository = BooksRepository(sessionManager: SessionManager.shared)
+        let viewModel = BooksTableViewModel(booksRepository: booksRepository)
+        let tableController = BooksTableViewController(viewModel: viewModel)
+        self.loadViewController(tableController, into: self._view.centralView)
```
Comment (danielaRiesgo) [inline WBooks/BookTab/HomeTabViewController.swift]: Remove `self`.

File: WBooks/BookTable/BooksTableViewController.swift [hunk 6]
```diff
@@ -33,6 +33,23 @@ final class BooksTableViewController: UIViewController {
...
+}
+
+extension BooksTableViewController {
+    
+    func bindViewModel() {
+        _viewModel.fetchBooks.startWithResult { [unowned self] result in
+            switch result {
+            case .Success: self._view.reloadData()
+            case .Failure(_):
+                let alert = UIAlertController(title: "Could not load books list", message: .None, preferredStyle: UIAlertControllerStyle.Alert)
```
Comment (danielaRiesgo) [inline WBooks/BookTable/BooksTableViewController.swift]: Localize string.

File: WBooks/BookTable/BooksTableViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+//  Copyright © 2017 Wolox. All rights reserved.
+//
+
+import Foundation
+import ReactiveCocoa
+
+class BooksTableViewModel {
+    
+    private var _viewModels: [BookViewModel] = []
+    let _booksRepository: BooksRepository
```
Comment (danielaRiesgo) [inline WBooks/BookTable/BooksTableViewModel.swift]: This should be `private` too.

File: WBooks/BookTable/BooksTableViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+extension BooksTableViewModel {
+    
+    var fetchBooks: SignalProducer<(), RepositoryError> {
+        return _booksRepository.fetchBooks()
+            .on(next: { [unowned self] books in self._viewModels = books.map { BookViewModel(book: $0) } })
+            .map { _ in return }
+    }
+    
+    var count: Int {
```
Comment (danielaRiesgo) [inline WBooks/BookTable/BooksTableViewModel.swift]: I think this and the subscript is not used (although it should in the table data source delegate).

File: WBooks/Login/LoginViewController.swift [hunk 9]
```diff
@@ -37,10 +37,15 @@ final class LoginViewController: UIViewController {
         GIDSignIn.sharedInstance().uiDelegate = self
         GIDSignIn.sharedInstance().delegate = self
         
-        let action = Action<Void, Void, NoError> { _ in
```
Comment (danielaRiesgo) [inline WBooks/Login/LoginViewController.swift]: Why did you undo this change?

File: WBooks/Repositories/BooksRepository.swift [hunk 10]
```diff
@@ -21,9 +21,9 @@ public class BooksRepository: BaseRepository {
         super.init(baseURL: baseURL, versionAPI: "v1", headers: headers, sessionManager: sessionManager)
     }
     
-    func fetchBooks() -> SignalProducer<BooksList, RepositoryError> {
-        return self.performRequest(Method.GET, path: BooksPath, parameters: .None) { JSON -> Result<BooksList, DecodeError> in
-            let result: Result<BooksList, DecodeError> = decode(JSON).toResult()
+    func fetchBooks() -> SignalProducer<[Book], RepositoryError> {
+        return self.performRequest(Method.GET, path: BooksPath, parameters: .None) { JSON -> Result<[Book], DecodeError> in
+            let result: Result<[Book], DecodeError> = decode(JSON).toResult()
```
Comment (danielaRiesgo) [inline WBooks/Repositories/BooksRepository.swift]: You could directly do:
`return decode(JSON).toResult()`
since the compiler can infer the type from the type declaration of the function.

File: WBooks/BookTab/HomeTabViewController.swift [hunk 11]
```diff
@@ -66,21 +62,10 @@ private extension HomeTabViewController {
...
-                    self.loadViewController(tableController, into: self._view.centralView)
-                case .Failure(_):
-                    let alert = UIAlertController(title: "Could not load books list", message: .None, preferredStyle: UIAlertControllerStyle.Alert)
-                    alert.addAction(UIAlertAction(title: "OK".localized(), style: UIAlertActionStyle.Default, handler: .None))
-                    self.presentViewController(alert, animated: true, completion: .None)
-                }
-            }
-
-        }
+        let booksRepository = BooksRepository(sessionManager: SessionManager.shared)
```
Comment (danielaRiesgo) [inline WBooks/BookTab/HomeTabViewController.swift]: This is a dependency. So the controller should receive it when it's created.

File: WBooks/BookTab/HomeTabViewController.swift [hunk 12]
```diff
@@ -61,9 +62,10 @@ private extension HomeTabViewController {
     func showLibrary() {
         navigationItem.title = "LIBRARY".localized().uppercaseString
         
-        let viewModel: BooksListViewModel = BooksListViewModel()
+        let booksRepository = BooksRepository(sessionManager: SessionManager.shared)
+        let viewModel = BooksTableViewModel(booksRepository: booksRepository)
         let tableController = BooksTableViewController(viewModel: viewModel)
-        loadViewController(tableController, into: _view.centralView)
+        self.loadViewController(tableController, into: self._view.centralView)
```
Comment (danielaRiesgo) [inline WBooks/BookTab/HomeTabViewController.swift]: Unnecessary `self`s.

File: WBooks/BookTable/BooksTableViewModel.swift (hunk 8 above)
Comment (danielaRiesgo) [inline WBooks/BookTable/BooksTableViewModel.swift]: @adrian-ponce-wolox This comment remains valid.

File: WBooks/BookTable/BooksTableViewController.swift [hunk 13]
```diff
@@ -33,6 +33,23 @@ final class BooksTableViewController: UIViewController {
...
+    
+}
+
+extension BooksTableViewController {
+    
+    func bindViewModel() {
+        _viewModel.fetchBooks.startWithResult { [unowned self] result in
+            switch result {
+            case .Success: self._view.reloadData()
+            case .Failure(_):
```
Comment (pablog) [inline WBooks/BookTable/BooksTableViewController.swift]: You can avoid the parameter: `case .Failure:`

File: WBooks/Login/LoginViewController.swift (hunk 9 above)
Comment (pablog) [inline WBooks/Login/LoginViewController.swift]: It was not working because `UIButton` cannot be casted to `Void`

File: WBooks/Repositories/BooksRepository.swift [hunk 14]
```diff
@@ -13,4 +13,19 @@ import Result
...
+    let BooksPath = "books"
+    
+    init(sessionManager: SessionManagerType) {
+        let baseURL: String = NSBundle.mainBundle()["BaseURL"]!
+        let headers = ["Content-Type": "application/json", "Accept": "application/json" ]
+        super.init(baseURL: baseURL, versionAPI: "v1", headers: headers, sessionManager: sessionManager)
+    }
+    
+    func fetchBooks() -> SignalProducer<[Book], RepositoryError> {
+        return self.performRequest(Method.GET, path: BooksPath, parameters: .None) { JSON -> Result<[Book], DecodeError> in
```
Comment (pablog) [inline WBooks/Repositories/BooksRepository.swift]: You can be less verbose:

```
return performRequest(.GET, path: BooksPath, parameters: .None) {
  return decode($0).toResult()
}
```

File: WBooks/Repositories/BooksRepository.swift [hunk 15]
```diff
@@ -13,4 +13,19 @@ import Result
 
 public class BooksRepository: BaseRepository {
     
+    let BooksPath = "books"
```
Comment (dtilve) [inline WBooks/Repositories/BooksRepository.swift]: `private static`

File: WBooks/BookTable/BooksTableViewController.swift [hunk 16]
```diff
@@ -33,6 +33,23 @@ final class BooksTableViewController: UIViewController {
         _view.delegate = self
         
         _view.registerCell(BookTableViewCell.self)
+        bindViewModel()
+    }
+    
+}
+
+extension BooksTableViewController {
```
Comment (dtilve) [inline WBooks/BookTable/BooksTableViewController.swift]: remember this extension should be `private` if this is only used within this controller

File: WBooks/BookTable/BooksTableViewModel.swift (hunk 8 above)
Comment (pablog) [inline WBooks/BookTable/BooksTableViewModel.swift]: What about this?

File: WBooks/Repositories/BooksRepository.swift [hunk 17]
```diff
@@ -22,9 +22,8 @@ public class BooksRepository: BaseRepository {
     }
     
     func fetchBooks() -> SignalProducer<[Book], RepositoryError> {
-        return self.performRequest(Method.GET, path: BooksPath, parameters: .None) { JSON -> Result<[Book], DecodeError> in
-            let result: Result<[Book], DecodeError> = decode(JSON).toResult()
-            return result
+        return self.performRequest(Method.GET, path: BooksRepository.BooksPath, parameters: .None) {
```
Comment (danielaRiesgo) [inline WBooks/Repositories/BooksRepository.swift]: Remove `self`.

File: WBooks/BookTable/BooksTableViewModel.swift (hunk 8 above)
Comment (adrian-ponce-wolox) [inline WBooks/BookTable/BooksTableViewModel.swift]: It is used on BooksTableViewController: UITableViewDataSource

File: WBooks/Login/LoginViewController.swift (hunk 9 above)
Comment (adrian-ponce-wolox) [inline WBooks/Login/LoginViewController.swift]: I have corrected it to: let action = Action<AnyObject, Void, NoError> { _ in
